File: scripts/generate-system-guide-pdf.mjs
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { jsPDF } from "jspdf";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const root = path.resolve(__dirname, "..");
const source = path.join(root, "docs", "bunnywell-portal-system-guide.md");
const output = path.join(root, "docs", "bunnywell-portal-system-guide.pdf");

const markdown = fs.readFileSync(source, "utf8");
const doc = new jsPDF({ unit: "pt", format: "a4" });

const page = {
  width: doc.internal.pageSize.getWidth(),
  height: doc.internal.pageSize.getHeight(),
  marginX: 48,
  marginTop: 54,
  marginBottom: 48,
};

let y = page.marginTop;
let pageNumber = 1;
let inCode = false;
let codeBuffer = [];

function stripMarkdown(text) {
  return text
    .replace(/\*\*([^*]+)\*\*/g, "$1")
    .replace(/`([^`]+)`/g, "$1")
    .replace(/\[(.*?)\]\((.*?)\)/g, "$1 ($2)");
}

function footer() {
  doc.setFont("helvetica", "normal");
  doc.setFontSize(8);
  doc.setTextColor("#617169");
  doc.text("Bunnywell Portal System Guide", page.marginX, page.height - 24);
  doc.text(`Page ${pageNumber}`, page.width - page.marginX, page.height - 24, { align: "right" });
}

function newPage() {
  footer();
  doc.addPage();
  pageNumber += 1;
  y = page.marginTop;
}

function ensureSpace(height) {
  if (y + height > page.height - page.marginBottom) newPage();
}

function writeWrapped(text, options = {}) {
  const {
    size = 10,
    font = "normal",
    color = "#24352E",
    indent = 0,
    before = 0,
    after = 6,
    lineHeight = size + 4,
    bullet = false,
  } = options;
  const maxWidth = page.width - page.marginX * 2 - indent;
  const cleanText = stripMarkdown(text);
  doc.setFont("helvetica", font);
  doc.setFontSize(size);
  doc.setTextColor(color);
  const lines = doc.splitTextToSize(cleanText, maxWidth - (bullet ? 12 : 0));
  ensureSpace(before + lines.length * lineHeight + after);
  y += before;
  if (bullet) {
    doc.text("-", page.marginX + indent, y);
    doc.text(lines, page.marginX + indent + 12, y);
  } else {
    doc.text(lines, page.marginX + indent, y);
  }
  y += lines.length * lineHeight + after;
}

function writeCode() {
  if (codeBuffer.length === 0) return;
  doc.setFont("courier", "normal");
  doc.setFontSize(8);
  doc.setTextColor("#24352E");
  const lineHeight = 11;
  const lines = codeBuffer.flatMap((line) => doc.splitTextToSize(line || " ", page.width - page.marginX * 2 - 16));
  ensureSpace(lines.length * lineHeight + 18);
  doc.setDrawColor("#D9DED6");
  doc.setFillColor("#FBFAF6");
  doc.roundedRect(page.marginX, y, page.width - page.marginX * 2, lines.length * lineHeight + 12, 4, 4, "FD");
  doc.text(lines, page.marginX + 8, y + 14);
  y += lines.length * lineHeight + 20;
  codeBuffer = [];
}

for (const rawLine of markdown.split(/\r?\n/)) {
  const line = rawLine.trimEnd();

  if (line.startsWith("```")) {
    if (inCode) writeCode();
    inCode = !inCode;
    continue;
  }

  if (inCode) {
    codeBuffer.push(line);
    continue;
  }

  if (!line.trim()) {
    y += 4;
    continue;
  }

  if (line.startsWith("# ")) {
    writeWrapped(line.slice(2), { size: 24, font: "bold", color: "#0F3D2E", after: 10, lineHeight: 29 });
  } else if (line.startsWith("## ")) {
    writeWrapped(line.slice(3), { size: 15, font: "bold", color: "#0F3D2E", before: 8, after: 5, lineHeight: 19 });
  } else if (line.startsWith("### ")) {
    writeWrapped(line.slice(4), { size: 12, font: "bold", color: "#0F3D2E", before: 5, after: 4, lineHeight: 16 });
  } else if (line.startsWith("- ")) {
    writeWrapped(line.slice(2), { size: 9.5, indent: 10, bullet: true, after: 3, lineHeight: 13 });
  } else if (/^\|.*\|$/.test(line)) {
    writeWrapped(line.replace(/\|/g, "  "), { size: 8.5, font: "normal", after: 3, lineHeight: 12 });
  } else {
    writeWrapped(line, { size: 9.5, after: 5, lineHeight: 13 });
  }
}

writeCode();
footer();
doc.save(output);
console.log(`Wrote ${output}`);

File: scripts/generate-system-guide-pdf.py
from pathlib import Path
import re

from reportlab.lib import colors
from reportlab.lib.pagesizes import A4
from reportlab.lib.styles import ParagraphStyle, getSampleStyleSheet
from reportlab.lib.units import mm
from reportlab.platypus import (
    SimpleDocTemplate,
    Paragraph,
    Spacer,
    Table,
    TableStyle,
    PageBreak,
)


ROOT = Path(__file__).resolve().parents[1]
SOURCE = ROOT / "docs" / "bunnywell-portal-system-guide.md"
OUTPUT = ROOT / "docs" / "bunnywell-portal-system-guide.pdf"


def escape(text: str) -> str:
    return (
        text.replace("&", "&amp;")
        .replace("<", "&lt;")
        .replace(">", "&gt;")
    )


def inline(text: str) -> str:
    text = escape(text)
    text = re.sub(r"`([^`]+)`", r"<font name='Courier'>\1</font>", text)
    text = re.sub(r"\*\*([^*]+)\*\*", r"<b>\1</b>", text)
    return text


def footer(canvas, doc):
    canvas.saveState()
    canvas.setFont("Helvetica", 8)
    canvas.setFillColor(colors.HexColor("#617169"))
    canvas.drawString(18 * mm, 12 * mm, "Bunnywell Portal System Guide")
    canvas.drawRightString(192 * mm, 12 * mm, f"Page {doc.page}")
    canvas.restoreState()


def build_story(markdown: str):
    styles = getSampleStyleSheet()
    body = ParagraphStyle(
        "Body",
        parent=styles["BodyText"],
        fontName="Helvetica",
        fontSize=9.5,
        leading=13,
        textColor=colors.HexColor("#24352E"),
        spaceAfter=5,
    )
    h1 = ParagraphStyle(
        "H1",
        parent=styles["Heading1"],
        fontName="Helvetica-Bold",
        fontSize=22,
        leading=26,
        textColor=colors.HexColor("#0F3D2E"),
        spaceAfter=10,
    )
    h2 = ParagraphStyle(
        "H2",
        parent=styles["Heading2"],
        fontName="Helvetica-Bold",
        fontSize=14,
        leading=18,
        textColor=colors.HexColor("#0F3D2E"),
        spaceBefore=8,
        spaceAfter=6,
    )
    bullet = ParagraphStyle(
        "Bullet",
        parent=body,
        leftIndent=12,
        firstLineIndent=-8,
        bulletIndent=0,
    )
    code = ParagraphStyle(
        "Code",
        parent=body,
        fontName="Courier",
        fontSize=8,
        leading=11,
        backColor=colors.HexColor("#F7F5EF"),
        borderPadding=4,
    )

    story = []
    in_table = False
    table_rows = []

    def flush_table():
      nonlocal table_rows, in_table
      if not table_rows:
          return
      table = Table(table_rows, colWidths=[28 * mm, 100 * mm, 38 * mm])
      table.setStyle(TableStyle([
          ("BACKGROUND", (0, 0), (-1, 0), colors.HexColor("#0F3D2E")),
          ("TEXTCOLOR", (0, 0), (-1, 0), colors.white),
          ("FONTNAME", (0, 0), (-1, 0), "Helvetica-Bold"),
          ("FONTNAME", (0, 1), (-1, -1), "Helvetica"),
          ("FONTSIZE", (0, 0), (-1, -1), 8),
          ("GRID", (0, 0), (-1, -1), 0.25, colors.HexColor("#D9DED6")),
          ("VALIGN", (0, 0), (-1, -1), "TOP"),
          ("ROWBACKGROUNDS", (0, 1), (-1, -1), [colors.white, colors.HexColor("#FBFAF6")]),
          ("LEFTPADDING", (0, 0), (-1, -1), 5),
          ("RIGHTPADDING", (0, 0), (-1, -1), 5),
      ]))
      story.append(table)
      story.append(Spacer(1, 6))
      table_rows = []
      in_table = False

    lines = markdown.splitlines()
    code_block = []
    in_code = False

    for raw in lines:
        line = raw.rstrip()
        if line.startswith("```"):
            if in_code:
                story.append(Paragraph("<br/>".join(escape(item) for item in code_block), code))
                code_block = []
                in_code = False
            else:
                flush_table()
                in_code = True
            continue
        if in_code:
            code_block.append(line)
            continue

        if line.startswith("|") and line.endswith("|"):
            cells = [Paragraph(inline(cell.strip()), body) for cell in line.strip("|").split("|")]
            if all(re.fullmatch(r"-+", cell.getPlainText().strip()) for cell in cells):
                continue
            table_rows.append(cells)
            in_table = True
            continue
        elif in_table:
            flush_table()

        if not line:
            story.append(Spacer(1, 4))
            continue
        if line.startswith("# "):
            story.append(Paragraph(inline(line[2:]), h1))
            story.append(Spacer(1, 4))
            continue
        if line.startswith("## "):
            story.append(Paragraph(inline(line[3:]), h2))
            continue
        if line.startswith("- "):
            story.append(Paragraph(inline(line[2:]), bullet, bulletText="-"))
            continue
        story.append(Paragraph(inline(line), body))

    flush_table()
    return story


def main():
    OUTPUT.parent.mkdir(parents=True, exist_ok=True)
    markdown = SOURCE.read_text(encoding="utf-8")
    doc = SimpleDocTemplate(
        str(OUTPUT),
        pagesize=A4,
        rightMargin=18 * mm,
        leftMargin=18 * mm,
        topMargin=18 * mm,
        bottomMargin=18 * mm,
        title="Bunnywell Portal System Guide",
    )
    doc.build(build_story(markdown), onFirstPage=footer, onLaterPages=footer)
    print(f"Wrote {OUTPUT}")


if __name__ == "__main__":
    main()

